sable_history/src/pg_history_service.rs

```diff
@@ -0,0 +1,315 @@
+                            .filter(messages::dsl::timestamp.lt(end_ts))
+                            // total order, consistent across requests
+                            .order((messages::dsl::timestamp, messages::dsl::id))
+                            .limit(limit)
+                            .load_stream(&mut *connection_lock),
+                    )
+                    .await
+                    .map_err(|e| {
+                        tracing::error!("Could not query messages: {e}");
+                        HistoryError::InternalError("Could not query messages".to_string())
+                    })?
+                    .map_ok(|row| make_historical_event(&channel, row))
+                    .try_collect::<Vec<_>>()
+                    .await
+                    .map_err(|e| {
+                        tracing::error!("Could not parse messages: {e}");
+                        HistoryError::InternalError("Could not parse message".to_string())
+                    })?
+                } else {
+                    let start_ts = DateTime::from_timestamp(start_ts, 0)
+                        .unwrap_or(DateTime::<Utc>::MAX_UTC)
+                        .naive_utc();
+                    let end_ts = DateTime::from_timestamp(end_ts, 999_999)
+                        .unwrap_or(DateTime::<Utc>::MIN_UTC)
+                        .naive_utc();
+                    let limit = i64::min(10000, i64::try_from(limit).unwrap_or(i64::MAX));
+                    Box::new(
+                        base_query
+                            .filter(messages::dsl::timestamp.gt(end_ts))
+                            .filter(messages::dsl::timestamp.lt(start_ts))
+                            // total order, consistent across requests
+                            .order((messages::dsl::timestamp.desc(), messages::dsl::id.desc()))
+                            .limit(limit)
+                            .load_stream(&mut *connection_lock),
+                    )
+                    .await
+                    .map_err(|e| {
+                        tracing::error!("Could not query messages: {e}");
+                        HistoryError::InternalError("Could not query messages".to_string())
+                    })?
+                    .map_ok(|row| make_historical_event(&channel, row))
+                    .try_collect::<Vec<_>>()
+                    .await
+                    .map_err(|e| {
+                        tracing::error!("Could not parse messages: {e}");
+                        HistoryError::InternalError("Could not parse message".to_string())
+                    })?
+                    .into_iter()
+                    .rev() // need to reverse *after* applying the SQL LIMIT
+                    .collect::<Vec<_>>()
+                }
+            }
+        })
+    }
+}
+
+fn make_historical_event(
+    channel: &crate::models::Channel,
+    (id, timestamp, message_type, text, source_nick, source_ident, source_vhost, source_account): (
```

Comment:
This might be more readable as a new Queryable struct (possibly with a helper method in the model to query it) - this ties into the comment further up about a helper query method

Reply:
I don't see what you have in mind